go/toolchains.rst

```diff
@@ -18,22 +27,568 @@ customize the behavior of the core_ go rules.
+need a `forked version of go`_, however you may want to `control the version`_ or use the
+`installed sdk`_.
+
+The toolchain
+~~~~~~~~~~~~~
+
+Declaration
+^^^^^^^^^^^
+
+Toolchains are declared using the go_toolchain_ macro. This actually registers two Bazel
+toolchains, the main go toolchain, and a special bootstrap toolchain. The bootstrap toolchain
+is needed because the full toolchain includes tools that are compiled on demand and written in
+go, so we need a special cut down version of the toolchain to build those tools.
+
+Toolchains are pre-declared for all the known combinations of host, target and sdk, and the names
+are a predictable
+go\ **version**\ _\ **host**
+for host toolchains and
+go\ **version**\ _\ **host**\ _cross\_\ **target**
+for cross compilation toolchains. So for instance if the rules_go repository is loaded with
+it's default name, the following toolchain labels (along with many others) will be available
+
+.. code::
+
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64-bootstrap
+  @io_bazel_rules_go//go/toolchain:1.9.0_linux_amd64_cross_windows_amd64
+
+The toolchains are not usable until you register_ them.
+
+Registration
+^^^^^^^^^^^^
+
+Normally you would just call go_register_toolchains_ from your WORKSPACE to register all the
+pre-declared toolchains, and allow normal selection logic to pick the right one.
+It is fine to add more toolchains to the set available if you like, and because the normal
+toolchain matching mechaism prefers the first declared match you can also override individual
+toolchains by having another one with the same constraints that you register *before* calling
+the go_register_toolchains_.
+
+If you wish to have more control over the toolchains you can instead just make direct
+calls to register_toolchains_ with only the toolchains you wish to install, you can see an
+example of this in `limiting the available toolchains`_.
+It is important to note that you **must** also register the boostrap toolchain for any other
+toolchain that you register, otherwise the tools for that toolchain cannot be built.
+
+Use
+^^^
+
+If you are writing a new rule that wants to use the go toolchain, you need to do a couple of things.
+First, you have to declare that you want to consume the toolchain on the rule declaration.
+
+.. code:: bzl
+
+  my_rule = rule(
+      _my_rule_impl,
+      attrs = {
+          ...
+      },
+      toolchains = ["@io_bazel_rules_go//go:toolchain"],
+  )
+
+And then in the rule body, you need to get the toolchain itself and use it's action generators.
+
+.. code:: bzl
+
+  def _my_rule_impl(ctx):
+    go_toolchain = ctx.toolchains["@io_bazel_rules_go//go:toolchain"]
+    srcs, vars = go_toolchain.actions.cover(ctx, go_toolchain, ctx.files.srcs)
+
+
+Customizing
+-----------
+
+Normal usage
+~~~~~~~~~~~~
+
+This is an example of normal usage for the other examples to be compared against.
+
```

Comment:
This will download and use the latest Go SDK that was available when the version of rules_go you're using was released.